src/Microsoft.Restier.Publishers.OData/RestierController.cs

```diff
@@ -525,6 +578,127 @@ private ODataPath GetPath()
+            var parameterArray = method.GetParameters();
+
+            var model = await Api.GetModelAsync(cancellationToken);
+
+            // Parameters of method and model is exactly mapped or there is parsing error
+            var parameters = new Collection<object>();
+
+            // Add binding parameter which is first parameter of method
+            var enumerableType = parameterArray[0].ParameterType.FindGenericType(typeof (IEnumerable<>));
+            if (enumerableType != null)
+            {
+                // This means function is bound to an entity set.
+                // IQueryable should always have generic type argument
+                var elementClrType = enumerableType.GenericTypeArguments[0];
+
+                // For entity set, user can write as ICollection<> or IEnumerable<> or array as method parameters
+                if (parameterArray[0].ParameterType.IsArray)
+                {
+                    var toListMethodInfo = typeof(Enumerable).GetMethod("ToArray").MakeGenericMethod(elementClrType);
+                    var listResult1 = toListMethodInfo.Invoke(null, new object[] { bindingParameterValue });
+                    parameters.Add(listResult1);
+                }
+                else if (parameterArray[0].ParameterType.FindGenericType(typeof (ICollection<>)) != null)
+                {
+                    var toListMethodInfo = typeof (Enumerable).GetMethod("ToList").MakeGenericMethod(elementClrType);
+                    var listResult = toListMethodInfo.Invoke(null, new object[] {bindingParameterValue});
+                    parameters.Add(listResult);
+                }
+                else
+                {
+                    parameters.Add(bindingParameterValue);
+                }
+            }
+            else
+            {
+                parameters.Add(bindingParameterValue.SingleOrDefault());
+            }
+
+            for (int i=1;i< parameterArray.Length;i++)
+            {
+                var parameter =parameterArray[i];
+                var currentParameterValue = segment.GetParameterValue(parameter.Name);
+                var parameterTypeRef = parameter.ParameterType.GetTypeReference(model);
+                var convertedValue = await DeserializationHelpers.ConvertValueAsync(currentParameterValue, parameter.ParameterType, parameterTypeRef, Api);
+                parameters.Add(convertedValue);
+            }
+
+            return await ExecuteFunction(method, parameters, cancellationToken);
+        }
+
+        private async Task<IQueryable> ExecuteFunction(MethodInfo method, Collection<object> parameters, CancellationToken cancellationToken)
+        {
+            object result = method.Invoke(Api, parameters.ToArray());
+            var model = await Api.GetModelAsync(cancellationToken);
+            IEdmTypeReference returnType = method.ReturnType.GetReturnTypeReference(model);
+            var asQueryableMethod = typeof(Queryable).GetMethod(
```

Comment:
For these "constant" instances, we should only compute them once and put them into a static class just like `Where`, `Select`, etc.